File: algorithms/graph_algorithms/bfs/bfs.js
// Breadth-First Search: Level-by-level exploration


function bfs_iterative(graph, start, goal) {
    if (start === goal) return [start];
    
    const queue = [[start, [start]]];
    const visited = new Set([start]);
    
    while (queue.length > 0) {
        const [node, path] = queue.shift();
        
        for (const neighbor of (graph[node] || [])) {
            if (visited.has(neighbor)) continue;
            if (neighbor === goal) return [...path, neighbor];
            visited.add(neighbor);
            queue.push([neighbor, [...path, neighbor]]);
        }
    }
    
    return null;
}


function bfs_level_order(graph, start) {
    const levels = [];
    const queue = [start];
    const visited = new Set([start]);
    
    while (queue.length > 0) {
        const levelSize = queue.length;
        const level = [];
        
        for (let i = 0; i < levelSize; i++) {
            const node = queue.shift();
            level.push(node);
            
            for (const neighbor of (graph[node] || [])) {
                if (!visited.has(neighbor)) {
                    visited.add(neighbor);
                    queue.push(neighbor);
                }
            }
        }
        levels.push(level);
    }
    
    return levels;
}


function bfs_bidirectional(graph, start, goal) {
    if (start === goal) return [start];
    
    let front = new Map([[start, [start]]]);
    let back = new Map([[goal, [goal]]]);
    let visitedFront = new Set([start]);
    let visitedBack = new Set([goal]);
    
    while (front.size > 0 && back.size > 0) {
        if (front.size > back.size) {
            [front, back] = [back, front];
            [visitedFront, visitedBack] = [visitedBack, visitedFront];
        }
        
        const nextFront = new Map();
        for (const [node, path] of front) {
            for (const neighbor of (graph[node] || [])) {
                if (visitedBack.has(neighbor)) {
                    return [...path, ...back.get(neighbor).reverse()];
                }
                if (!visitedFront.has(neighbor)) {
                    visitedFront.add(neighbor);
                    nextFront.set(neighbor, [...path, neighbor]);
                }
            }
        }
        
        front = nextFront;
    }
    
    return null;
}


module.exports = { bfs_iterative, bfs_level_order, bfs_bidirectional };

File: algorithms/graph_algorithms/bfs/bfs.rb
# Breadth-First Search: Level-by-level exploration
require 'set'


def bfs_iterative(graph, start, goal)
  return [start] if start == goal
  
  queue = [[start, [start]]]
  visited = Set[start]
  
  until queue.empty?
    node, path = queue.shift
    (graph[node] || []).each do |neighbor|
      next if visited.include?(neighbor)
      return path + [neighbor] if neighbor == goal
      visited.add(neighbor)
      queue << [neighbor, path + [neighbor]]
    end
  end
  
  nil
end


def bfs_level_order(graph, start)
  levels = []
  queue = [start]
  visited = Set[start]
  
  until queue.empty?
    level_size = queue.size
    level = []
    level_size.times do
      node = queue.shift
      level << node
      (graph[node] || []).each do |neighbor|
        unless visited.include?(neighbor)
          visited.add(neighbor)
          queue << neighbor
        end
      end
    end
    levels << level
  end
  
  levels
end


def bfs_bidirectional(graph, start, goal)
  return [start] if start == goal
  
  front = { start => [start] }
  back = { goal => [goal] }
  visited_front = Set[start]
  visited_back = Set[goal]
  
  until front.empty? || back.empty?
    front, back = back, front if front.size > back.size
    visited_front, visited_back = visited_back, visited_front if front.size > back.size
    
    next_front = {}
    front.each do |node, path|
      (graph[node] || []).each do |neighbor|
        return path + back[neighbor].reverse if visited_back.include?(neighbor)
        unless visited_front.include?(neighbor)
          visited_front.add(neighbor)
          next_front[neighbor] = path + [neighbor]
        end
      end
    end
    
    front = next_front
  end
  
  nil
end
